[3.x] feat: add NoConstructorDependencyInjectionRule for console commands

ggiak opened the pull request:
## Description

  This PR adds a new rule that enforces Laravel's best practice of using method injection in the `handle()` method rather than constructor injection for console commands.

  ### Background

  Laravel's documentation and community best practices recommend avoiding constructor dependency injection in console commands. Instead, dependencies should be injected into the `handle()`
  method. This pattern:
  - Improves testability
  - Follows Laravel conventions
  - Avoids potential issues with command initialization
  - Makes commands more consistent with the framework

  ### What this rule does

  - Detects console commands extending `Illuminate\Console\Command`
  - Checks if the constructor has any parameters
  - Reports violations with helpful error messages and tips
  - Is disabled by default for backward compatibility

  ### Example

  ```php
  // This will trigger an error:
  class MyCommand extends Command
  {
      public function __construct(private UserService $userService)
      {
          parent::__construct();
      }
  }

  // Error: Console command "MyCommand" should not have constructor arguments. Use dependency injection in the handle() method instead.
```

 ### Configuration

To enable the rule, add to your phpstan.neon:
  parameters:
      noConsoleCommandConstructorInjection: true

- Added or updated tests
- Documented user facing changes

###  Changes

- Added new rule NoConstructorDependencyInjectionRule in src/Rules/ConsoleCommand/
- Added comprehensive test coverage with various scenarios (commands with/without constructors, abstract commands, etc.)
- Added documentation in docs/rules.md explaining the rule, its purpose, and configuration
- Updated extension.neon to register the rule with opt-in configuration parameter

###  Breaking changes

None. The rule is disabled by default to maintain backward compatibility. Users must explicitly enable it in their configuration.

ggiak committed:
feat: add NoConstructorDependencyInjectionRule for console commands

This rule enforces Laravel's best practice of using method injection in the handle() method
rather than constructor injection for console commands. The rule:

- Detects console commands extending Illuminate\Console\Command
- Checks if the constructor has any parameters
- Reports violations with helpful error messages and tips
- Is disabled by default for backward compatibility

To enable the rule, add to your phpstan.neon:
```
parameters:
    noConsoleCommandConstructorInjection: true
```

Co-authored-by: zoispag <zoispag@users.noreply.github.com>

canvural commented on src/Rules/ConsoleCommand/NoConstructorDependencyInjectionRule.php:
```diff
@@ -0,0 +1,79 @@
+{
+    public function getNodeType(): string
+    {
+        return InClassNode::class;
```
Why don't we hook into `InClassMethod` node? And then we can also do something like
```php
$method = $node->getMethodReflection();

if (!$method->isConstructor()) {
	return [];
}
```

in `processNode` method.

ggiak replied:
Great point. 

added an additional commit that uses

InClassMethodNode instead.

https://github.com/larastan/larastan/pull/2324/commits/87f9219ed17cd8c5ce673d06d18da0e123e88e34

calebdw commented on src/Rules/ConsoleCommand/NoConstructorDependencyInjectionRule.php:
```diff
@@ -0,0 +1,79 @@
+            return [
+                RuleErrorBuilder::message(
+                    sprintf(
+                        'Console command "%s" should not have constructor arguments. Use dependency injection in the handle() method instead.',
```
Commands can also be invokable (use `__invoke` method instead of `handle`)

szepeviktor replied:
We could be strict and allow only 1 way (`./artisan make:command`) to implement a command.

canvural replied:
No. We can be dynamic here and suggest `__invoke` or `handle` depending which one the class has.

ggiak replied:
@calebdw not a great fan of __invoke but you are correct.

@szepeviktor not our job to be opinionated 👍  

@canvural 
As per making the message dynamic, I would suggest to simply change the message instead of identifying the used method. 
Reasoning: People change preferences over time and it's good to raise awareness that someone can use both.

In this spirit, I din't change the logic but I changed the corresponding message to:

1) Console command "%s" should not have constructor arguments. Use dependency injection in the handle() or __invoke() method instead.

2) Tip: "Move all dependencies to the handle() or __invoke() method parameters for better testability and Laravel best practices."

However, I do not have a strong opinion on the matter so feel free to suggest otherwise  and I'd be happy to apply the changes.

ggiak committed:
refactor: improve NoConstructorDependencyInjectionRule implementation

- Refactor rule to use InClassMethodNode instead of InClassNode for better performance
- Use isConstructor() method for cleaner constructor detection
- Add support for invokable commands using __invoke() method
- Update error messages to mention both handle() and __invoke() methods
- Fix test namespaces to follow consistent pattern (Tests\Rules\ConsoleCommand)
- Add test cases for invokable commands

This improves the rule's efficiency by only processing method nodes instead of entire classes,
and extends support to Laravel's invokable command pattern.

calebdw commented on docs/rules.md:
```diff
@@ -559,3 +559,59 @@ parameters:
 
+## NoConstructorDependencyInjectionRule
+
+This rule ensures that console commands follow Laravel's best practice of using method injection in the `handle()` method rather than constructor injection.
```
docs should be updated to reflect the __invoke method as well

calebdw commented on src/Rules/ConsoleCommand/NoConstructorDependencyInjectionRule.php:
```diff
@@ -0,0 +1,67 @@
+
+        $classReflection = $node->getClassReflection();
+
+        if (! $classReflection->isSubclassOf(Command::class)) {
```
`isSubclassOf` with a class string is deprecated, this should be:


```suggestion
        if (! $classReflection->is(Command::class)) {
```

calebdw commented on src/Rules/ConsoleCommand/NoConstructorDependencyInjectionRule.php:
```diff
@@ -0,0 +1,67 @@
+                    ),
+                )->line($methodNode->getLine())
+                    ->tip('Move all dependencies to the handle() or __invoke() method parameters for better testability and Laravel best practices.')
+                    ->identifier('larastan.console.constructorInjection')
```
generally, I've only ever seen two components to the identifier, I would recommend:


```suggestion
                    ->identifier('larastan.consoleConstructorInjection')
```

ggiak replied:
there are a bunch:

- src/Rules/DeferrableServiceProviderMissingProvidesRule.php:55
- src/Rules/UselessConstructs/NoUselessValueFunctionCallsRule.php:51

and more, but the most important one is the:
- src/Rules/ConsoleCommand/UndefinedArgumentOrOptionRule.php:78

which resides at the same namespace level as the newly proposed rule.
for the sake of continuety, I would suggest to leave this one out.

wdty ?

calebdw replied:
I would still say those are the exception and not the rule, but it's up to you

ggiak replied:
Then I am performing the change accordingly and I will create a corresponding issue with "help wanted"

ggiak replied:
@calebdw https://github.com/larastan/larastan/pull/2324/commits/5874cac31b32fb54f0aed5dc8ee0f3dbf9f86757

calebdw commented on src/Rules/ConsoleCommand/NoConstructorDependencyInjectionRule.php:
```diff
@@ -0,0 +1,67 @@
+                        $classReflection->getName(),
+                    ),
+                )->line($methodNode->getLine())
+                    ->tip('Move all dependencies to the handle() or __invoke() method parameters for better testability and Laravel best practices.')
```
you're saying the same thing in the message and the tip---I would recommend shortening the actual message. Additionally, I would shorten the tip:


```suggestion
                    sprintf(
                        'Console command "%s" should not have constructor arguments.',
                        $classReflection->getName(),
                    ),
                )->line($methodNode->getLine())
                    ->tip('Move all dependencies to the handle() or __invoke() methods.')
```

calebdw reviewed (approved):
Looks pretty good, just a few last nits

ggiak committed:
Code review changes :2

ggiak commented:
> Looks pretty good, just a few last nits

Thank you for the prompt reply / check

Latest changes: https://github.com/larastan/larastan/pull/2324/commits/7cd8b4ed01eccc984964f98069685d4e8764115b

Let me know about: [larastan.console.constructorInjection](https://github.com/search?q=repo%3Alarastan%2Flarastan+identifier%28%27larastan.&type=code)

ggiak committed:
changing larastan.console.constructorInjection to larastan.consoleConstructorInjection

ggiak commented:
@canvural mind if you check this one? it's an important rule. 

When dependency injection goes to the constructor then the dependencies gets initiailized during the application load.
